### DISCUSSION TITLE
Workaround for trait with associated types and lifetimes

### DISCUSSION THREAD

**Discussion by dsomerfield-vendia** [asomers/mockall#575] [Questions]
I have a fairly complex case where I have to cast some associated types that have lifetimes. I'm not terribly surprised that automock can't figure it out, but I also haven't been able to figure out if there is a work-around. This is the an example of the situation that I can't mock:

```
#[async_trait]
trait Repo {
    type QueryEngine: QueryEngine;

    async fn run_query(&self, query: &<Self::QueryEngine as QueryEngine>::Query<'_>) -> Result<(), ()>;
}
```

I have tried to do a manual mock, but haven't been able to create it in a way that the mock! macro will accept. Here is the mock I tried to create:

```
    mock! {
        Repo {}

        #[async_trait]
        impl Repo for Repo {
            type QueryEngine = MockQueryEngine;

            async fn run_query(&self, query: &<Self::QueryEngine as QueryEngine>::Query<'_>) -> Result<(), ()>;

        }

    }
```

When I manually mock, it doesn't seem to like code generated by mock!, specifically, the `Self` reference (`ambiguous associated type`) and the inferred lifetime (`missing lifetime specifier`). Is there anyway, I can provide some hints, or possibly make some assumption that will allow the generated code to be acceptable. Full example with full error message is [here](https://gist.github.com/dsomerfield-vendia/24457ea418ff8fdb56e83f52e280933a).

Incidentally, the example is async, but it I have removed all async code and that doesn't seem to be affecting anything at this point.

**Comment by dsomerfield-vendia** [comment]
It figures. After fighting with this for hours, I might have just figured it out. This manual mock seems to work:

```
    mock! {
        Repo {}

        impl Repo for Repo {
            type QueryEngine = MockQueryEngine;
            fn run_query<'a>(&self, query: &<<MockRepo as Repo>::QueryEngine as QueryEngine>::Query<'a>) -> Result<(), ()>;
        }

    }
```

I need to validate whether that is semantically the same as the inferred lifetime, but this gives me a path forward. I

**Comment by asomers** [comment]
Issue #541 describes the same anonymous lifetime problem, and shows a workaround.  As for `Self`, you could try replacing it with `MockRepo`.

**Reply by dsomerfield-vendia** [reply]
You are so right on both counts. Thank you! I was very close, but this nudged me to the solution.

BTW, if you are looking for contributors to this project, let me know. I'm becoming a very dedicated user and would love to give back.
